src/block_cache/block_chunkcache.c

```diff
@@ -738,6 +738,78 @@ __wt_chunkcache_remove(
     return (0);
 }
 
+/*
+ * __wt_chunkcache_ingest --
+ *     Read all the contents from a file and insert it into the chunkcache.
+ */
+int
+__wt_chunkcache_ingest(
+  WT_SESSION_IMPL *session, const char *local_name, const char *sp_obj_name, uint32_t object_id)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_CHUNKCACHE_CHUNK *chunk;
+    WT_CHUNKCACHE_HASHID hash_id;
+    WT_DECL_RET;
+    WT_FH *fh;
+    size_t already_read, size;
+    uint64_t bucket_id;
+
+    chunkcache = &S2C(session)->chunkcache;
+    already_read = 0;
+
+    WT_RET(__wt_open(session, local_name, WT_FS_OPEN_FILE_TYPE_DATA, WT_FS_OPEN_READONLY, &fh));
+    WT_ERR(__wt_filesize(session, fh, (wt_off_t *)&size));
+
+    while (already_read < size) {
+        bucket_id = __chunkcache_tmp_hash(
+          chunkcache, &hash_id, sp_obj_name, object_id, (wt_off_t)already_read);
+
+        __wt_spin_lock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+
+        if ((ret = __chunkcache_alloc_chunk(
+               session, (wt_off_t)already_read, (wt_off_t)size, &hash_id, &chunk)) != 0) {
+            __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+            WT_ERR(ret);
+        }
+
+        /*
+         * Insert the invalid chunk into the bucket before releasing the lock and doing I/O. This
+         * way we avoid two threads trying to cache the same chunk.
+         */
+        TAILQ_INSERT_HEAD(WT_BUCKET_CHUNKS(chunkcache, bucket_id), chunk, next_chunk);
+        __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+
+        /* Read the new chunk. Only one thread would be caching the new chunk. */
+        if ((ret = __wt_read(
+               session, fh, chunk->chunk_offset, chunk->chunk_size, chunk->chunk_memory)) != 0) {
+            __wt_spin_lock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+            TAILQ_REMOVE(WT_BUCKET_CHUNKS(chunkcache, bucket_id), chunk, next_chunk);
+            __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+            __chunkcache_free_chunk(session, chunk);
+            WT_STAT_CONN_INCR(session, chunk_cache_io_failed);
+            return (ret);
+        }
+
+        /*
+         * Mark chunk as valid. The only thread that could be executing this code is the thread that
+         * won the race and inserted this (invalid) chunk into the hash table. This thread has now
+         * read the chunk, while any other threads that were looking for the same chunk would be
+         * spin-waiting for this chunk to become valid. The current thread will mark the chunk as
+         * valid, and any waiters will unblock and proceed reading it.
+         */
+        WT_PUBLISH(chunk->valid, true);
```

Comment:
Seeing this concurrency code repeated strengthens the case for factoring this out. But I understand you're already working on it.